01_pangram.py

```diff
@@ -0,0 +1,193 @@
+      # variable somewhere?
+      # * if calculated every time, then o(n) linear time to calculate the l
+      # length of the set.
+      # space - o(1)?
+      # * probably constant, just a single comparison?
+      # * may not need to go through the whole phrase if the phrase is a
+      # pangram and all alphabets are reached before the phrase ends
+      if len(alpha) == 0:
+        return True
+
+    return False
+
+  def is_pangram_2(self, phrase):
+    """
+    Determines whether a given string is a pangram
+    Time analysis references https://wiki.python.org/moin/TimeComplexity
+    where possible
+
+    Overall
+    * time - o(n). `for c in phrase` runs a max of n times where n is length
+     of `phrase`
+    * space - o(1) cos everything in the function requires constant space
+    * although if we take storage of `phrase` into account then o(n) where n
+    * is length of the phrase
+
+    Function may terminate early if `phrase` is a pangram and all alphabets are
+     reached before the phrase ends
+
+    The only difference from is_pangram_1 is that `len(alpha) == 0` becomes
+    `alpha == set()`
+    """
+    # time - o(1)
+    # space - o(1)
+    # o(26) constant space and time, since the number of ascii lowercase
+    # characters is constant - 26. so o(1) * 26 = o(26)
+    # which simplifies back to o(1)
+    alpha = set('abcdefghijklmnopqrstuvwxyz')
+
+    # overall loop
+    # time - o(n)
+    # * o(n) `for c in phrase`
+    # space - o(1)?
+    # * just a fixed number of temporary variables that are ints or chars?
+    # alpha doesn't get any bigger, always smaller or the same size
+
+    # `c in phrase`
+    # time - o(n).
+    # * is o(1) done max n times, so o(n) where n is the number of characters
+    # in `phrase`
+    # space - o(1)? constant?
+    # * since c is a temporary variable
+    for c in phrase:
+      # o(1) constant time and space?
+      # does set.discard take constant time and space?
+      alpha.discard(c.lower())
+
+      # o(1) time and space probably - just a single comparison?
+      # may not need to go through the whole phrase if the phrase is a pangram
+      # and all alphabets are reached before the phrase ends
+      if alpha == set():
+        return True
+
+    return False
+
+  def is_pangram_3(self, phrase):
+    """
+    Determines whether a given string is a pangram
+    Time analysis references https://wiki.python.org/moin/TimeComplexity
+    where possible
+
+    Overall
+    * time - o(n) where n is `phrase`'s length
+    * space - o(n) where n is `phrase`'s length
+    * both due to the time and space to convert `phrase` into a set in
+    * `set(phrase.lower)`
+    """
+    # time - o(1)
+    # space - o(1)
+    # since the number of ascii lowercase characters is constant - 26
+    # so o(1) * 26 which simplifies back to constant o(1)
+    alpha = set('abcdefghijklmnopqrstuvwxyz')
+
+    # overall
+    # time - o(n)
+    # space - o(n)
+    # where n is the length of `phrase`
```

Comment:
Is this o(n) total or o(n) not including the input?

Reply:
@robot-dreams my understanding is o(n) total where n is length of `phrase` because a larger phrase may result in a larger set. (eg: 500-character phrase where each char is unique takes up more space than a 10-character phrase where each char is unique)